### ISSUE TITLE
Assembly based entries

### ISSUE THREAD

**Issue by timbussmann** [ParticularLabs/NServiceBus.FileBasedRouting#2]
The current routing file requires one entry per message type. Compared to MessageEndpointMappings, this can become quite extensive to maintain. We should consider adding support for assembly and namespace based definitions, e.g.

```
<handles>
    <!-- current entry -->
    <command type="Commands.MyCommand, SomeOtherAssembly" />
    <!-- assembly based entries -->
    <commands assembly="Contracts" namespace="Commands" />
    <commands assembly="MyEndpoint />
</handles>
```

thoughts?

**Comment by timbussmann** [comment]
compared to message endpoint mappings, we would have the chance of forming a clearer API here, e.g for a single type use `<command ... />` while for "collections" use <commands .../>?

/cc @DavidBoike

**Comment by timbussmann** [comment]
also, completely different structures may be a bit cleaner from a syntax perspective, e.g.

```
<handles>
    <commands>
        <assembly name="CommandsAssembly">
            <namespace name="commands1">
                <type name="Command1" />
            </namespace>
            <namespace name="commands2">
                <all />
            </namespace>
        </assembly>
    </commands>
    <events>
        <assembly name="EventsAssembly">
            <all />
        </assembly>
    </events>
</handles
```

but that quickly becomes quite exhaustive to write.

Other ideas? @SzymonPobiega @Scooletz

**Comment by SzymonPobiega** [comment]
I like the attribute-based syntax more. 👍 for adding namespace/assembly.

Also, not sure about the events vs commands distinction. I know @andreasohlund and @SimonCropp had this crazy idea of making the conventions optional. Would having a distinction hare, in the routing config, interfere with making the conventions optional?

**Comment by timbussmann** [comment]
> Also, not sure about the events vs commands distinction

that's a good point. I feel that being more clear about the differnce would help. Also especially when we want to add validation logic to the file (e.g. check a command is only handled by single endpoint, whereas events can be handled by multiple endpoints).

> this crazy idea of making the conventions optional. Would having a distinction hare, in the routing config, interfere with making the conventions optional?

I think theoretically it would also work by registering every single type found when using the assembly based approaches, as we have different routing tables for send and publish. The only downside without conventions would be, that those tables would be much larger as we couldn't filter out message types. But otherwise I think that would work with or without conventions.

**Comment by SzymonPobiega** [comment]
After thinking about it once more, I agree that having a distinction on the routing level does not imply we have it in conventions so should be good. Worst case we'll filter only based on `IsMessage`.

**Comment by timbussmann** [comment]
addressed by #5
